Mathlib/Combinatorics/Enumerative/InclusionExclusion.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2024 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.BigOperators.Pi
+import Mathlib.Algebra.BigOperators.Ring
+import Mathlib.Algebra.Module.BigOperators
+
+/-!
+# Inclusion-exclusion principle
+
+This file proves several variants of the inclusion-exclusion principle.
+
+## TODO
+
+Use `Finset.prod_indicator_biUnion_sub_indicator` to prove the integral version of
+inclusion-exclusion.
+-/
+
+open Int
+
+namespace Finset
+variable {ι α G : Type*} [DecidableEq α] [AddCommGroup G] {s : Finset ι}
+
+lemma prod_indicator_biUnion_sub_indicator (hs : s.Nonempty) (f : ι → Finset α) (a : α) :
+    ∏ i ∈ s, (Set.indicator (s.biUnion f) 1 a - Set.indicator (f i) 1 a) = (0 : ℤ) := by
+  by_cases ha : a ∈ s.biUnion f
+  · obtain ⟨i, hi, ha⟩ := mem_biUnion.1 ha
+    exact prod_eq_zero hi <| by simp [*, -coe_biUnion]
+  · obtain ⟨i, hi⟩ := hs
+    have ha : a ∉ f i := fun h ↦ ha <| subset_biUnion_of_mem _ hi h
+    exact prod_eq_zero hi <| by simp [*, -coe_biUnion]
+
+/-- **Inclusion-exclusion principle** for the sum of a function over a union.
+
+The sum of a function `g` over the union of the `f i` over `i ∈ s` is the alternating sum of the
+sums of `g` over the intersections of the `f i`. -/
+theorem inclusion_exclusion_sum_biUnion (s : Finset ι) (f : ι → Finset α) (g : α → G) :
```

Comment:
I have argued in the past that the symbol-reading convention is suboptimal for named theorems because
1. There are a lot of symbols to mention
2. Variants of the same named theorem might have very different symbols, meaning that
  a. The name of one variant won't help you guess the name of the others
  b. The names make it hard to know how exactly the variants differ from each other

Here, it's easy to see at a glance that the four `theorem`s are variants of the same named theorem, and the names directly tell you whether it's talking about the `sum` or the `card`, and whether it's stated using `biUnion` or `Finset.inf _ fun _ ↦ _ᶜ`

Reply:
OK, I see your point and I agree.